typed_python/Instance.hpp

```diff
@@ -0,0 +1,166 @@
+#pragma once
```

Comment:
if you break this into .cc and .h, the .h can fwd declare Type and won't need to include Type.hpp.

Reply:
I'll give that a shot. 

Separately, I tried to break `Type.hpp` to multiple files but got stuck because of the `check` method which requires knowing its child classes. If you think there is a way around that we can look into it when you're back next week